diff --git a/tools/cmake/file_downloader.py b/tools/cmake/file_downloader.py
index 729c64e..0376e8c 100644
--- a/tools/cmake/file_downloader.py
+++ b/tools/cmake/file_downloader.py
@@ -263,15 +263,23 @@ def download_extract_files(items):
             }
             if "user-agent" in item:
                 headers["User-Agent"] = item["user-agent"]
-            err_times = 3
+            err_times = 5
             delay_time = 0
             while 1:
+                failed = False
                 try:
                     d = Downloader(item["url"], pkg_path, max_thead_num = 8, force_write=True, headers=headers)
                     d.start()
                     break
                 except Exception as e:
                     print("-- download failed:", e)
+                    failed = True
+                if "sha256sum" in item and item["sha256sum"] and not failed:
+                    ok, sha256sum = check_sha256sum(pkg_path, item["sha256sum"])
+                    failed = not ok
+                if failed:
+                    if os.path.exists(pkg_path):
+                        os.remove(pkg_path)
                     err_times -= 1
                     if err_times > 0:
                         delay_time += 5
