From 695646ae819e740c90a762228055add816ad3144 Mon Sep 17 00:00:00 2001
From: scpcom <scpcom@gmx.de>
Date: Fri, 10 Oct 2025 20:56:42 +0200
Subject: [PATCH] server: hid: lock before reset

---
 server/service/hid/reset.go | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/server/service/hid/reset.go b/server/service/hid/reset.go
index c9b883e..4cdbffa 100644
--- a/server/service/hid/reset.go
+++ b/server/service/hid/reset.go
@@ -12,6 +12,14 @@ import (
 func (s *Service) Reset(c *gin.Context) {
 	var rsp proto.Response
 
+	h := GetHid()
+	h.Lock()
+	h.CloseNoLock()
+	defer func() {
+		h.OpenNoLock()
+		h.Unlock()
+	}()
+
 	// reset USB
 	f, err := os.OpenFile("/sys/kernel/config/usb_gadget/g0/UDC", os.O_WRONLY, 0644)
 	if err != nil {
-- 
2.34.1

