From 66bc16c38b929f59667aa5289b5b9a2af8fa6c95 Mon Sep 17 00:00:00 2001
From: scpcom <scpcom@gmx.de>
Date: Sun, 12 Oct 2025 00:36:25 +0200
Subject: [PATCH] server: hid: set mode without reboot

---
 server/service/hid/status.go | 225 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 219 insertions(+), 6 deletions(-)

diff --git a/server/service/hid/status.go b/server/service/hid/status.go
index 9a77cb9..5cbb76c 100644
--- a/server/service/hid/status.go
+++ b/server/service/hid/status.go
@@ -16,7 +15,15 @@ import (
 const (
 	ModeNormal  = "normal"
 	ModeHidOnly = "hid-only"
-	ModeFlag    = "/sys/kernel/config/usb_gadget/g0/bcdDevice"
+	ModeKbdOnly = "kbd-only"
+
+	ModeNonBios = "normal"
+	ModeHidBios = "bios"
+
+	HidGadgetPath     = "/sys/kernel/config/usb_gadget/g0"
+	HidConfPath       = "/sys/kernel/config/usb_gadget/g0/configs/c.1"
+	HidFuncPath       = "/sys/kernel/config/usb_gadget/g0/functions"
+	ModeFlag          = "/sys/kernel/config/usb_gadget/g0/bcdDevice"
 
 	ModeNormalScript  = "/kvmapp/system/init.d/S03usbdev"
 	ModeHidOnlyScript = "/kvmapp/system/init.d/S03usbhid"
@@ -26,6 +32,12 @@ const (
 var modeMap = map[string]string{
 	"0x0510": ModeNormal,
 	"0x0623": ModeHidOnly,
+	"0x0102": ModeKbdOnly,
+}
+
+var biosMap = map[string]string{
+	"0": ModeNonBios,
+	"1": ModeHidBios,
 }
 
 func (s *Service) GetHidMode(c *gin.Context) {
@@ -48,14 +60,6 @@ func (s *Service) SetHidMode(c *gin.Context) {
 		return
 	}
 
-	h := GetHid()
-	h.Lock()
-	h.CloseNoLock()
-	defer func() {
-		h.OpenNoLock()
-		h.Unlock()
-	}()
-
 	srcScript := ModeNormalScript
 	if req.Mode == ModeHidOnly {
 		srcScript = ModeHidOnlyScript
@@ -71,11 +83,15 @@ func (s *Service) SetHidMode(c *gin.Context) {
 		return
 	}
 
-	rsp.OkRsp(c)
+	biosmode, _ := getBiosMode();
 
-	log.Println("reboot system...")
-	time.Sleep(500 * time.Millisecond)
-	_ = exec.Command("reboot").Run()
+	msg, err := setHidMode(req.Mode, biosmode)
+	if err != nil {
+		rsp.ErrRsp(c, -4, msg)
+		return
+	}
+
+	rsp.OkRsp(c)
 }
 
 func disable() (error) {
@@ -156,3 +172,200 @@ func getHidMode() (string, error) {
 
 	return mode, nil
 }
+
+func setHidMode(hidmode, biosmode string) (string, error) {
+	h := GetHid()
+	h.Lock()
+	h.CloseNoLock()
+	defer func() {
+		h.OpenNoLock()
+		h.Unlock()
+	}()
+
+	if err := disable(); err != nil {
+		log.Errorf("disable hid failed: %s", err)
+		return "disable hid failed", err
+	}
+
+	flag := "0"
+	if biosmode == ModeHidBios {
+		flag = "1"
+	}
+
+	for _, hidfunc := range []string{"hid.GS0", "hid.GS1", "hid.GS2"} {
+		hidFunction := fmt.Sprintf("%s/%s", HidFuncPath, hidfunc)
+		hidSymLink := fmt.Sprintf("%s/%s", HidConfPath, hidfunc)
+		hidBiosFlag := fmt.Sprintf("%s/subclass", hidFunction)
+
+		if err := os.Remove(hidSymLink); err != nil {
+			log.Fatalf("Could not remove symlink: %v", err)
+			return "unlink hid failed", err
+		}
+
+		// bios mode
+		if err := os.WriteFile(hidBiosFlag, []byte(flag), 0o666); err != nil {
+			log.Errorf("set bios mode failed: %s", err)
+			return "set bios mode failed", err
+		}
+
+		if err := os.Symlink(hidFunction, hidSymLink); err != nil {
+			log.Errorf("Could not create symlink: %v", err)
+			return "symlink hid failed", err
+		}
+	}
+
+	usbFlag := "0x0200"
+	devFlag := "0x0510"
+	idVen := "0x3346"
+	idPrd := "0x1009"
+	maxPkt := "0x40"
+	strVen := "sipeed"
+	strPrd := "NanoKVM"
+	if hidmode == ModeHidOnly {
+		usbFlag = "0x0101"
+		devFlag = "0x0623"
+	} else if hidmode == ModeKbdOnly {
+		usbFlag = "0x0001"
+		devFlag = "0x0102"
+		idVen = "0x05ac"
+		idPrd = "0x0202"
+		maxPkt = "0x08"
+		strVen = "Alps Electric"
+		strPrd = "Apple USB Keyboard"
+        }
+
+	hidBcdUsb := fmt.Sprintf("%s/%s", HidGadgetPath, "bcdUSB")
+	hidBcdDev := fmt.Sprintf("%s/%s", HidGadgetPath, "bcdDevice")
+	hidIdVen := fmt.Sprintf("%s/%s", HidGadgetPath, "idVendor")
+	hidIdPrd := fmt.Sprintf("%s/%s", HidGadgetPath, "idProduct")
+	hidMaxPkt := fmt.Sprintf("%s/%s", HidGadgetPath, "bMaxPacketSize0")
+	hidStrVen := fmt.Sprintf("%s/%s", HidGadgetPath, "strings/0x409/manufacturer")
+	hidStrPrd := fmt.Sprintf("%s/%s", HidGadgetPath, "strings/0x409/product")
+
+	if err := os.WriteFile(hidBcdUsb, []byte(usbFlag), 0o666); err != nil {
+		log.Errorf("set bcdUSB failed: %s", err)
+		return "set bcdUSB failed", err
+	}
+
+	if err := os.WriteFile(hidBcdDev, []byte(devFlag), 0o666); err != nil {
+		log.Errorf("set bcdDevice failed: %s", err)
+		return "set bcdDevice failed", err
+	}
+
+	if err := os.WriteFile(hidIdVen, []byte(idVen), 0o666); err != nil {
+		log.Errorf("set idVendor failed: %s", err)
+		return "set idVendor failed", err
+	}
+
+	if err := os.WriteFile(hidIdPrd, []byte(idPrd), 0o666); err != nil {
+		log.Errorf("set idProduct failed: %s", err)
+		return "set idProduct failed", err
+	}
+
+	if err := os.WriteFile(hidMaxPkt, []byte(maxPkt), 0o666); err != nil {
+		log.Errorf("set bMaxPacketSize0 failed: %s", err)
+		return "set bMaxPacketSize0 failed", err
+	}
+
+	if err := os.WriteFile(hidStrVen, []byte(strVen), 0o666); err != nil {
+		log.Errorf("set strings/0x409/manufacturer failed: %s", err)
+		return "set strings/0x409/manufacturer failed", err
+	}
+
+	if err := os.WriteFile(hidStrPrd, []byte(strPrd), 0o666); err != nil {
+		log.Errorf("set strings/0x409/product failed: %s", err)
+		return "set strings/0x409/product failed", err
+	}
+
+	bmAttr := "0xE0"
+	maxPwr := "120"
+	strCnf := "NanoKVM"
+	if hidmode == ModeHidOnly {
+		bmAttr = "0xA0"
+		maxPwr = "200"
+	} else if hidmode == ModeKbdOnly {
+		bmAttr = "0xA0"
+		maxPwr = "50"
+		strCnf = ""
+	}
+
+	hidBmAttr := fmt.Sprintf("%s/%s", HidConfPath, "bmAttributes")
+	hidMaxPwr := fmt.Sprintf("%s/%s", HidConfPath, "MaxPower")
+	hidStrCnf := fmt.Sprintf("%s/%s", HidConfPath, "strings/0x409/configuration")
+
+	if err := os.WriteFile(hidBmAttr, []byte(bmAttr), 0o666); err != nil {
+		log.Errorf("set bmAttributes failed: %s", err)
+		return "set bmAttributes failed", err
+	}
+
+	if err := os.WriteFile(hidMaxPwr, []byte(maxPwr), 0o666); err != nil {
+		log.Errorf("set MaxPower failed: %s", err)
+		return "set MaxPower failed", err
+	}
+
+	if err := os.WriteFile(hidStrCnf, []byte(strCnf), 0o666); err != nil {
+		log.Errorf("set strings/0x409/configuration failed: %s", err)
+		return "set strings/0x409/configuration failed", err
+	}
+
+	for _, hidfunc := range []string{"mass_storage.disk0", "ncm.usb0", "rndis.usb0"} {
+		hidFunction := fmt.Sprintf("%s/%s", HidFuncPath, hidfunc)
+		hidSymLink := fmt.Sprintf("%s/%s", HidConfPath, hidfunc)
+		hidFuncOn, _ := isFuncExist(hidFunction)
+		hidConfOn, _ := isFuncExist(hidSymLink)
+
+		if hidFuncOn && hidConfOn && hidmode != ModeNormal {
+			if err := os.Remove(hidSymLink); err != nil {
+				log.Fatalf("Could not remove symlink: %v", err)
+				return "unlink hid failed", err
+			}
+		} else if hidFuncOn && !hidConfOn && hidmode == ModeNormal {
+			if err := os.Symlink(hidFunction, hidSymLink); err != nil {
+				log.Errorf("Could not create symlink: %v", err)
+				return "symlink hid failed", err
+			}
+		}
+	}
+
+	if err := enable(); err != nil {
+		log.Errorf("enable hid failed: %s", err)
+		return "enable hid failed", err
+	}
+
+	return "", nil
+}
+
+func getBiosMode() (string, error) {
+	hidFunction := fmt.Sprintf("%s/%s", HidConfPath, "hid.GS0")
+	hidBiosFlag := fmt.Sprintf("%s/subclass", hidFunction)
+
+	data, err := os.ReadFile(hidBiosFlag)
+	if err != nil {
+		log.Errorf("failed to read %s: %s", hidBiosFlag, err)
+		return "", err
+	}
+
+	key := strings.TrimSpace(string(data))
+	mode, ok := biosMap[key]
+	if !ok {
+		log.Errorf("invalid bios flag: %s", key)
+		return "", errors.New("invalid bios flag")
+	}
+
+	return mode, nil
+}
+
+func isFuncExist(device string) (bool, error) {
+	_, err := os.Stat(device)
+
+	if err == nil {
+		return true, nil
+	}
+
+	if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	}
+
+	log.Errorf("check file %s err: %s", device, err)
+	return false, err
+}
diff --git a/web/src/i18n/locales/de.ts b/web/src/i18n/locales/de.ts
index 4f46ffb..e460c72 100644
--- a/web/src/i18n/locales/de.ts
+++ b/web/src/i18n/locales/de.ts
@@ -94,7 +94,7 @@ const de = {
         desc: 'Wenn Ihre Maus und Tastatur nicht mehr reagieren und das Zurücksetzen der HID-Verbindung nicht hilft, könnte es sich um ein Kompatibilitätsproblem zwischen dem NanoKVM und dem Gerät handeln. Versuchen Sie, den HID-Only Modus zu aktivieren, um die Kompatibilität zu verbessern.',
         tip1: 'Die Aktivierung des HID-Only Modus entfernt das virtuelle U-Laufwerk und das virtuelle Netzwerk.',
         tip2: 'Im HID-Only Modus ist das Einbinden von Systemabbilder deaktiviert.',
-        tip3: 'NanoKVM wird nach dem Wechsel in den neuen Modus automatisch neu gestartet.',
+        tip3: 'NanoKVM wird nach dem Wechsel in den neuen Modus HID zurücksetzen.',
         enable: 'HID-Only Modus aktivieren',
         disable: 'HID-Only Modus deaktivieren'
       }
diff --git a/web/src/i18n/locales/en.ts b/web/src/i18n/locales/en.ts
index fe0cec3..b7fd4fe 100644
--- a/web/src/i18n/locales/en.ts
+++ b/web/src/i18n/locales/en.ts
@@ -98,7 +98,7 @@ const en = {
         desc: "If your mouse and keyboard stop responding and resetting HID doesn't help, it could be a compatibility issue between the NanoKVM and the device. Try to enable HID-Only mode for better compatibility.",
         tip1: 'Enabling HID-Only mode will unmount the virtual U-disk and virtual network',
         tip2: 'In HID-Only mode, image mounting is disabled',
-        tip3: 'NanoKVM will automatically reboot after switching modes',
+        tip3: 'NanoKVM will reset HID after switching modes',
         enable: 'Enable HID-Only mode',
         disable: 'Disable HID-Only mode'
       }
-- 
2.34.1

