From e200e05daaa392febda73d22b70ebd9dba5f7e04 Mon Sep 17 00:00:00 2001
From: scpcom <scpcom@gmx.de>
Date: Sun, 12 Oct 2025 19:12:49 +0200
Subject: [PATCH] server: hid: mode: check if func exists

---
 server/service/hid/mode.go | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/server/service/hid/mode.go b/server/service/hid/mode.go
index 5cbb76c..b49d25d 100644
--- a/server/service/hid/mode.go
+++ b/server/service/hid/mode.go
@@ -196,10 +196,14 @@ func setHidMode(hidmode, biosmode string) (string, error) {
 		hidFunction := fmt.Sprintf("%s/%s", HidFuncPath, hidfunc)
 		hidSymLink := fmt.Sprintf("%s/%s", HidConfPath, hidfunc)
 		hidBiosFlag := fmt.Sprintf("%s/subclass", hidFunction)
+		hidFuncOn, _ := isFuncExist(hidFunction)
+		hidConfOn, _ := isFuncExist(hidSymLink)
 
-		if err := os.Remove(hidSymLink); err != nil {
-			log.Fatalf("Could not remove symlink: %v", err)
-			return "unlink hid failed", err
+		if hidConfOn {
+			if err := os.Remove(hidSymLink); err != nil {
+				log.Fatalf("Could not remove symlink: %v", err)
+				return "unlink hid failed", err
+			}
 		}
 
 		// bios mode
@@ -208,9 +212,11 @@ func setHidMode(hidmode, biosmode string) (string, error) {
 			return "set bios mode failed", err
 		}
 
-		if err := os.Symlink(hidFunction, hidSymLink); err != nil {
-			log.Errorf("Could not create symlink: %v", err)
-			return "symlink hid failed", err
+		if hidFuncOn {
+			if err := os.Symlink(hidFunction, hidSymLink); err != nil {
+				log.Errorf("Could not create symlink: %v", err)
+				return "symlink hid failed", err
+			}
 		}
 	}
 
-- 
2.34.1

