From 427669c1b3e4ac3f13177f4957526fd07fdeb334 Mon Sep 17 00:00:00 2001
From: scpcom <scpcom@gmx.de>
Date: Tue, 14 Oct 2025 02:13:19 +0200
Subject: [PATCH] server: hid: mode: set otg role

---
 server/service/hid/status.go | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/server/service/hid/status.go b/server/service/hid/status.go
index 8936b45..56c23cd 100644
--- a/server/service/hid/status.go
+++ b/server/service/hid/status.go
@@ -20,6 +20,7 @@ const (
 	ModeNonBios = "normal"
 	ModeHidBios = "bios"
 
+	CviUsbOtgRole     = "/proc/cviusb/otg_role"
 	HidGadgetPath     = "/sys/kernel/config/usb_gadget/g0"
 	HidConfPath       = "/sys/kernel/config/usb_gadget/g0/configs/c.1"
 	HidFuncPath       = "/sys/kernel/config/usb_gadget/g0/functions"
@@ -510,6 +512,10 @@ func setHidMode(hidmode, biosmode string) (string, error) {
 		return "enable hid failed", err
 	}
 
+	if msg, err := setOtgRole(true); err != nil {
+		return msg, err
+	}
+
 	return "", nil
 }
 
@@ -544,6 +550,27 @@ func getBiosMode() (string, error) {
 	return mode, nil
 }
 
+func setOtgRole(devicemode bool) (string, error) {
+	isCviUsb, _ := isFuncExist(CviUsbOtgRole)
+
+	// do not fail on non-cvi hardware
+	if !isCviUsb {
+		return "", nil;
+	}
+
+	otgRole := "device"
+	if !devicemode {
+		otgRole = "host"
+	}
+
+	if err := os.WriteFile(CviUsbOtgRole, []byte(otgRole), 0o666); err != nil {
+		log.Errorf("set otg role failed: %s", err)
+		return "set otg role failed", err
+	}
+
+	return "", nil;
+}
+
 func isFuncExist(device string) (bool, error) {
 	_, err := os.Stat(device)
 
-- 
2.34.1

