From 8b5eea73f1ef405559bce6fa45a54bbcc4416196 Mon Sep 17 00:00:00 2001
From: scpcom <scpcom@gmx.de>
Date: Tue, 14 Oct 2025 02:59:31 +0200
Subject: [PATCH] server: hid: disable/enable: use HidGadgetPath

---
 server/service/hid/status.go | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/server/service/hid/status.go b/server/service/hid/status.go
index 7eb064a..ae7a19e 100644
--- a/server/service/hid/status.go
+++ b/server/service/hid/status.go
@@ -38,6 +39,7 @@ func (s *Service) Reset(c *gin.Context) {
 }
 
 func disable() (error) {
+	hidGadgetUDC := fmt.Sprintf("%s/%s", HidGadgetPath, "UDC")
 	hidGadgetOn, _ := isFuncExist(HidGadgetPath)
 
 	// gadget already disabled
@@ -46,15 +48,15 @@ func disable() (error) {
 	}
 
 	// reset USB
-	f, err := os.OpenFile("/sys/kernel/config/usb_gadget/g0/UDC", os.O_WRONLY, 0644)
+	f, err := os.OpenFile(hidGadgetUDC, os.O_WRONLY, 0644)
 	if err != nil {
-		log.Errorf("open /sys/kernel/config/usb_gadget/g0/UDC failed: %s", err)
+		log.Errorf("open %s failed: %s", hidGadgetUDC, err)
 		return err
 	}
 	err = f.Truncate(0)
 	if err != nil {
 		_ = f.Close()
-		log.Errorf("truncate /sys/kernel/config/usb_gadget/g0/UDC failed: %s", err)
+		log.Errorf("truncate %s failed: %s", hidGadgetUDC, err)
 		return err
 	}
 	_, err = f.Seek(0, 0)
@@ -66,7 +68,7 @@ func disable() (error) {
 	_, err = f.WriteString("\n")
 	if err != nil {
 		_ = f.Close()
-		log.Errorf("write to /sys/kernel/config/usb_gadget/g0/UDC failed: %s", err)
+		log.Errorf("write to %s failed: %s", hidGadgetUDC, err)
 		return err
 	}
 	_ = f.Close()
@@ -77,22 +79,24 @@ func disable() (error) {
 }
 
 func enable() (error) {
+	hidGadgetUDC := fmt.Sprintf("%s/%s", HidGadgetPath, "UDC")
+
 	devices, err := os.ReadDir("/sys/class/udc/")
 	if err != nil {
 		log.Errorf("read udc directory failed: %s", err)
 		return err
 	}
 
-	f, err := os.OpenFile("/sys/kernel/config/usb_gadget/g0/UDC", os.O_WRONLY, 0644)
+	f, err := os.OpenFile(hidGadgetUDC, os.O_WRONLY, 0644)
 	if err != nil {
-		log.Errorf("open /sys/kernel/config/usb_gadget/g0/UDC failed: %s", err)
+		log.Errorf("open %s failed: %s", hidGadgetUDC, err)
 		return err
 	}
 	for _, device := range devices {
 		_, err = f.WriteString(device.Name() + "\n")
 		if err != nil {
 			_ = f.Close()
-			log.Errorf("write to /sys/kernel/config/usb_gadget/g0/UDC failed: %s", err)
+			log.Errorf("write to %s failed: %s", hidGadgetUDC, err)
 			return err
 		}
 	}
-- 
2.34.1

