From b157b9fe28a90e0d152c6cb64b86ad4019ce2dc3 Mon Sep 17 00:00:00 2001
From: scpcom <scpcom@gmx.de>
Date: Mon, 1 Dec 2025 23:34:17 +0100
Subject: [PATCH] S03usbdev: fix virtual media split merge

---
 kvmapp/system/init.d/S03usbdev | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/kvmapp/system/init.d/S03usbdev b/kvmapp/system/init.d/S03usbdev
index c84dd94..23216ec 100755
--- a/kvmapp/system/init.d/S03usbdev
+++ b/kvmapp/system/init.d/S03usbdev
@@ -109,17 +109,20 @@ start_usb_dev(){
             disk=$(cat /boot/usb.disk0)
             if [ -z "${disk}" ]
             then
-                    # if [ ! -e /mnt/usbdisk.img ]
-                    # then
-                    #         fallocate -l 8G /mnt/usbdisk.img
-                    #         mkfs.vfat /mnt/usbdisk.img
-                    # fi
-                    echo /dev/mmcblk0p3 > functions/mass_storage.disk0/lun.0/file
+                    echo "" > functions/mass_storage.disk0/lun.0/file
             else
                     cat /boot/usb.disk0 > functions/mass_storage.disk0/lun.0/file
             fi
     fi
 
+    # mount data partition
+    if [ -e /boot/usb.disk1 ]
+    then
+            mkdir functions/mass_storage.disk1
+            ln -s functions/mass_storage.disk1 configs/c.1/
+            echo /dev/mmcblk0p3 > functions/mass_storage.disk1/lun.0/file
+    fi
+
     if [ ! -e /boot/udc.disable ]; then
         ls /sys/class/udc/ | cat > UDC
     fi
-- 
2.34.1

